File: src/main/scala/fi/oph/koski/schema/Oppija.scala
package fi.oph.koski.schema

import fi.oph.scalaschema.annotation._

@Description("Oppija Koski-järjestelmässä. Sisältää henkilötiedot ja listan opiskeluoikeuksista, jotka puolestaan sisältävät suorituksia, läsnäolotietoja jne. Henkilötietoja ei tallenneta Koskeen, vaan haetaan/tallennetaan Opintopolun henkilöpalveluun.")
case class Oppija(
  henkilö: Henkilö,
  @Description("Lista henkilön opiskeluoikeuksista. Sisältää vain ne opiskeluoikeudet, joihin käyttäjällä on oikeudet. Esimerkiksi ammatilliselle toimijalle ei välttämättä näy henkilön lukio-opintojen tietoja")
  opiskeluoikeudet: Seq[Opiskeluoikeus]
) {
  def tallennettavatOpiskeluoikeudet = opiskeluoikeudet.collect { case oo: KoskeenTallennettavaOpiskeluoikeus => oo }
  def ulkoisistaJärjestelmistäHaetutOpiskeluoikeudet = opiskeluoikeudet.diff(tallennettavatOpiskeluoikeudet)
}

File: web/app/types/fi/oph/koski/schema/Oppija.ts
import { Henkilö } from './Henkilo'
import { Opiskeluoikeus } from './Opiskeluoikeus'

/**
 * Oppija Koski-järjestelmässä. Sisältää henkilötiedot ja listan opiskeluoikeuksista, jotka puolestaan sisältävät suorituksia, läsnäolotietoja jne. Henkilötietoja ei tallenneta Koskeen, vaan haetaan/tallennetaan Opintopolun henkilöpalveluun
 *
 * @see `fi.oph.koski.schema.Oppija`
 */
export type Oppija = {
  $class: 'fi.oph.koski.schema.Oppija'
  henkilö: Henkilö
  opiskeluoikeudet: Array<Opiskeluoikeus>
}

export const Oppija = (o: {
  henkilö: Henkilö
  opiskeluoikeudet?: Array<Opiskeluoikeus>
}): Oppija => ({
  $class: 'fi.oph.koski.schema.Oppija',
  opiskeluoikeudet: [],
  ...o
})

export const isOppija = (a: any): a is Oppija =>
  a?.$class === 'fi.oph.koski.schema.Oppija'
